warehouse/accounts/forms.py

```diff
@@ -30,19 +30,67 @@ class LoginForm(forms.Form):
         ],
     )
 
-    def __init__(self, *args, login_service, **kwargs):
+    def __init__(self, *args, user_service, **kwargs):
         super().__init__(*args, **kwargs)
+        self.user_service = user_service
+
+
+class RegistrationForm(CredentialsMixin, forms.Form):
+    password_confirm = wtforms.PasswordField(
+        validators=[
+            wtforms.validators.DataRequired(),
+            wtforms.validators.EqualTo(
+                "password", "Passwords must match."
+            ),
+        ],
+    )
+
+    full_name = wtforms.StringField()
+
+    email = wtforms.fields.html5.EmailField(
+        validators=[
+            wtforms.validators.DataRequired(),
+            wtforms.validators.Email(),
+        ],
+    )
+
+    g_recaptcha_response = wtforms.StringField()
+
+    def __init__(self, *args, recaptcha_service, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.recaptcha_service = recaptcha_service
+
+    def validate_username(self, field):
+        if self.user_service.find_userid(field.data) is not None:
+            raise wtforms.validators.ValidationError(
+                "Username exists.")
+
+    def validate_email(self, field):
+        if self.user_service.find_userid_by_email(field.data) is not None:
+            raise wtforms.validators.ValidationError("Email exists.")
+
+    def validate_g_recaptcha_response(self, field):
+        # do required data validation here due to enabled flag being required
+        if self.recaptcha_service.enabled and not field.data:
+            raise wtforms.validators.ValidationError("Recaptcha error.")
+
+        try:
+            self.recaptcha_service.verify_response(field.data)
+        except recaptcha.RecaptchaError:
+            # TODO: log error
```

Comment:
If this is a "user did not type the recaptcha correctly" style error, then we probably don't want to log anything here (or at the most, we'd want debug level logging, not error level). Our errors get logged to sentry and in general errors should be actionable for us/operators.